docs(benchmarks): add Performance and Sizing Guide for record encryption (#3957)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/performance/con-about-performance-guide.adoc

////
  // Copyright Kroxylicious Authors.
  //
  // Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
////

:_mod-docs-content-type: CONCEPT

// file included in the following:
//
// performance-guide/index.adoc

[id='con-about-performance-guide-{context}']
= About this guide

[role="_abstract"]
This guide helps operators estimate the performance impact of deploying Kroxylicious as a proxy layer in front of Apache Kafka.

> It covers two configurations: a passthrough proxy with no filters enabled, and a proxy with the record encryption filter.

Reviewer: ```suggestion
It covers two configurations: a passthrough proxy with no filters enabled and a proxy with the record encryption filter.
```

> The goal is to give you the data you need to right-size your deployment: how much additional latency to expect, and where the throughput ceiling sits.

Reviewer: ```suggestion
The goal is to help you right-size your deployment by showing the expected latency overhead and throughput limits for each configuration.
```

== Summary

> Passthrough proxy (no filters)::

Reviewer: ```suggestion
Passthrough proxy (no filters enabled)::
```

> Negligible overhead.

Reviewer: Is this subjective? 
Maybe just stick to the figure in the next sentence

> Average publish latency increased by approximately 0.2 ms (+6%), with no measurable throughput impact.

Reviewer: Should the summary mention these measurements were returned under certain test conditions?

> The proxy layer itself is not a bottleneck.

Reviewer: Reads as an absolute statement under any conditions.
Maybe soften: "No proxy bottleneck was observed in the tested configurations."

> Record encryption (AES-256-GCM)::

Reviewer: ```suggestion
Record Encryption filter (AES-256-GCM)::
```

Reviewer: we also mention `Vault KMS` in the methodology section

> Measurable but predictable overhead.

Reviewer: `predictable` across hardware, workloads etc?

+

> * *Throughput:* approximately 26% reduction per partition (approximately 37,000 msg/s versus approximately 50,000 msg/s baseline with 1 KB messages).

Reviewer: ```suggestion
* *Throughput:* approximately 26% lower throughput per partition (37,000 msg/s compared to a baseline of 50,000 msg/s with 1 KB messages).
```

Reviewer: I think the first _approximately_ does all the work

* *Latency at sub-saturation rates:* average publish latency increased by 0.2–3 ms depending on load; p99 increased by 15–40 ms.
* *CPU cost:* encryption adds approximately 33% CPU overhead (13% direct crypto, 20% buffer management, GC, and re-encoding), consistent with the throughput reduction.

> * *Scaling:* throughput scales linearly with CPU allocation — approximately 40,000 msg/s per core at 1 KB. Validated at 1, 2, and 4 cores. Set `requests == limits` for a predictable, stable thread count.

Reviewer: ```suggestion
* *Scaling:* in the tested configurations, throughput scaled linearly with CPU allocation, reaching approximately 40,000 msg/s per core with 1 KB messages. Validated at 1, 2, and 4 cores. Set `requests == limits` for a predictable, stable thread count.
```
